dts/arm/nordic/nrf54l15_cpuapp.dtsi

```diff
@@ -55,6 +55,15 @@ nvic: &cpuapp_nvic {};
 		nordic,tasks-mask = <0x0003f800>;
 		status = "disabled";
 	};
+
+	egpio: gpio {
+		compatible = "nordic,nrf-egpio";
+		gpio-controller;
+		#gpio-cells = < 0x2 >;
+		ngpios = < 0x10 >;
+		status = "disabled";
+		port = < 0x2 >;
+	};
```

Comment:
I'm wondering if this is the right approach.
For advanced usecases we will need to use snippets anyway - enabling a certain emulated peripheral will cause other HW peripherals to be reserved. Similarly with messaging nodes, they also may vary depending on eperiph.
Listing all available ePeriphs here might be unnecessary.

Reply:
I agree. I think eperiphs should be included in the project through snippets, not listed in generic .dtsi files.

To be more precise, I think an emulated peripheral should be included in a project as a folder containing:

* out-of-tree device driver for the peripheral
* snippet for required Kconfig and dts modifications
* binary or source code creating a binary running in VPR

Reply:
Why keep the APP side driver out of tree though?
For customer specific ones it makes sense, but ones developed and maintained by us I would prefer to keep in upstream / at least in NCS.

Reply:
My thinking was that the development and delivery of all modules required for an emulated peripheral could be self-contained in a "package". An application developer could download this package and enable it in the project as simply as adding the snippet to the build.

I see an alternative would be to include the whole package in-tree: create an in-tree driver, keep required snippets in https://github.com/zephyrproject-rtos/zephyr/tree/main/snippets, include the VPR code as a binary or source code potentially from hal_nordic. This way also adding the peripheral to the application would be as simple as adding the snippet. I see this could have lower maintenance than out-of-tree components. However, the release process would be more complicated, because it would require upstreaming all of the components.

Another alternative would be to split the package and keep some parts in-tree (device driver, dts bindings?) and another part out-of-tree (VPR code, snippet?). The main cost of this approach is version tracking and compatibility management between the two parts.

Do you have a preference for the solution to choose?

Reply:
> An application developer could download this package and enable it in the project as simply as adding the snippet to the build.

This approach seems to assume that the FLPR firmware is delivered as a hex file, right?

>I see an alternative would be to include the whole package in-tree: create an in-tree driver, keep required snippets in https://github.com/zephyrproject-rtos/zephyr/tree/main/snippets, include the VPR code as a binary or source code potentially from hal_nordic.

Yes, that would be an alternative I see if we want to keep everything upstream. That's the preferable choice from my side as SHIM will be actively maintained there. The issue is where to place the FLPR code - it could either be a sample in zephyr or an application in hal_nordic (that would be something new though - we would need to think through if that is acceptable).

Another option is to keep both SHIM and FLPR code in NCS repo - SHIM would live in drivers and FLPR code in applications.

Or one more - keep FLPR code as source code in NCS's apps, build it as hex and include in zephyr's hal_nordic, and maintain SHIM in upstream zephyr.

Reply:
> This approach seems to assume that the FLPR firmware is delivered as a hex file, right?

Not necessarily. The package could include source code, but then it would also require a west manifest (or a similar solution) to point to the compatible NCS/Zephyr version.
Anyway, out-of-tree drivers also would need to somehow track compatibility with NCS/Zephyr. So that's probably undesired. If we wanted to keep the source code hidden, it would be probably better to include it in NCS as a srcx repository.

> Another option is to keep both SHIM and FLPR code in NCS repo - SHIM would live in drivers and FLPR code in applications.

I like this option because it allows for keeping compatibility between the Zephyr driver and the FLPR code easily. Moreover, it would allow for maintaining FLPR code aligned with the approach described in https://nordicsemi.atlassian.net/browse/NCSDK-10551. And for the closed-source FLPR applications, it would be possible to use already mentioned https://nordicsemi.atlassian.net/wiki/spaces/NCS/pages/108257943/srcx+repositories.

I'm not sure if there is any value in keeping emulated peripherals upstream.

Thoughts?

Reply:
The main value I see in keeping SHIM upstream (and having SHIM requires us to at least provide a hex so whole solution works) would be the fact that all changes introduced in Zephyr are applied to that SHIM without involvement on our side. So the cost of maintenance decreases.

Reply:
But I don't have a strong feeling about this. Keeping everything in-tree in NCS is more than fine for me.
@magp-nordic @jaz1-nordic what do you think?

Reply:
> The main value I see in keeping SHIM upstream (and having SHIM requires us to at least provide a hex so whole solution works) would be the fact that all changes introduced in Zephyr are applied to that SHIM without involvement on our side. So the cost of maintenance decreases.

Right. So it basically makes upmerges simpler. It could reduce Vestavind's workload needed to adjust downstream code to upstream changes.

In my opinion, the flexibility in protecting the IPR we have downstream is more valuable than making upmerges simpler. But it's just an opinion. I have no data supporting it.

Reply:
> But I don't have a strong feeling about this. Keeping everything in-tree in NCS is more than fine for me.
@magp-nordic @jaz1-nordic what do you think?

I think all code should be in one repo, both SHIM and FLPR. The main reason is the ease of recreating an emulated peripheral by someone outside of Nordic. If SHIM was in upstream Zephyr and FLPR code in NCS, then not only would someone need to navigate between two repos in terms of code but also follow their versions: for example, if the latest NCS version has the latest Zephyr, as Hubert mentioned. And I don't think this is the best user experience.
I also think that it might be difficult for us to maintain it due to too many dependencies. If we had SHIM in upstream, FLPR code in NCS, and FLPR hex in hal_nordic, then we would have to develop code in NCS(FLPR) before Zephyr release, so that we could include hex in hal_nordic on time. And what if upmerge comes along in the meantime or a freeze? It would be a mess.
So there are two options: keeping all code in Zephyr or in NCS. When it comes to Zephyr, I am not entirely sure it will be accepted by Zephyr community, especially FLPR code. It seems too Nordic-specific. Also, keeping FLPR code in Zephyr as a sample, so that anyone outside of Nordic could change our code, which should most probably be fixed/frozen after finishing, seems a bit risky.
So, to sum up, in my opinion keeping SHIM and FLPR codes in NCS is the best option.

Reply:
> But I don't have a strong feeling about this. Keeping everything in-tree in NCS is more than fine for me. @magp-nordic @jaz1-nordic what do you think?

Taking into account all the pros and cons, and the fact that the emulated peripherals on flpr are a Nordic-specific product, I am also in favor of keeping everything in one repo, which is NCS.

Reply:
Good, then we are all in agreement.
Let's keep both SHIMs and FLPR code in sdk-nrf.

Details:
When it comes to SHIMs I think it's pretty obvious they should live in `drivers` (and relevant snippets in `snippets`).
FLPR FW could be placed then:
- for closed source peripherals (if any), and ones that we prebuild and certify - in `srcx` repo.
- for open source peripherals - either in `applications` (seems to be a bit of an overshoot) or in `core_fw` (that is going to be established according to JIRA that Hubert linked).

@hubertmis @nika-nordic is this okay for you?

Reply:
> @hubertmis @nika-nordic is this okay for you?

Sounds perfect. It is aligned with the strategy for other software components in NCS.

Reply:
>Do I understand correctly that the snippet will be placed in nrf/snippets?

Yes, that would be a fitting place for them.

>Is it possible that the snippet would add an application or core_fw build to the project so that the project author wouldn't need to do any other actions than adding the snippet?

We need to look into whether this is possible with current implementation of snippets and sysbuild. But that would be our end goal.